models/classes/runner/navigation/QtiRunnerNavigation.php

```diff
@@ -101,41 +108,72 @@ public static function move($direction, $scope, RunnerServiceContext $context, $
-                $isJumpOutOfSection = false;
-                if (($nextPosition >= 0) && ($nextPosition < $route->count())) {
-                    $nextSection = $route->getRouteItemAt($nextPosition);
-
-                    // phpcs:disable Generic.Files.LineLength
-                    $isJumpOutOfSection = ($section->getIdentifier() !== $nextSection->getAssessmentSection()->getIdentifier());
-                    // phpcs:enable Generic.Files.LineLength
-                }
 
-                if ($isJumpOutOfSection && $limits != null && $limits->hasMaxTime()) {
-                    $assessmentItemRefs = $section->getComponentsByClassName('assessmentItemRef');
-                    foreach ($assessmentItemRefs as $assessmentItemRef) {
-                        $itemSessions = $session->getAssessmentItemSessions($assessmentItemRef->getIdentifier());
-
-                        if ($itemSessions !== false) {
-                            foreach ($itemSessions as $itemSession) {
-                                $itemSession->endItemSession();
-                            }
-                        }
-                    }
+            // As we have only one identifier for the whole adaptive section
+            // it will consider a jump of section on the first item
+            if (
+                !self::isAdaptive($context)
+                && self::jumpsOutOfSection($route, $section, $nextPosition)
+                && self::isTimeLimited($section)
+            ) {
+                self::endItemSessions($session, $section);
+            }
+        }
+    }
+
+    private static function endItemSessions(
+        AssessmentTestSession $session,
+        AssessmentSection $section
+    ): void {
+        foreach ($section->getComponentsByClassName('assessmentItemRef') as $assessmentItemRef) {
+            $itemSessions = $session->getAssessmentItemSessions($assessmentItemRef->getIdentifier());
+
+            if ($itemSessions !== false) {
+                foreach ($itemSessions as $itemSession) {
+                    $itemSession->endItemSession();
                 }
             }
         }
     }
+
+    private static function jumpsOutOfSection(
+        Route $route,
+        AssessmentSection $section,
+        $nextPosition
+    ): bool {
+        if (($nextPosition >= 0) && ($nextPosition < $route->count())) {
+            $nextSection = $route->getRouteItemAt($nextPosition);
+            $nextSectionId = $nextSection->getAssessmentSection()->getIdentifier();
+
+            return ($section->getIdentifier() !== $nextSectionId);
+        }
+
+        return false;
+    }
+
+    private static function isTimeLimited(AssessmentSection $section): bool
+    {
+        $limits = $section->getTimeLimits();
+
+        return ($limits != null) && $limits->hasMaxTime();
+    }
+
+    private static function isAdaptive(RunnerServiceContext $context): bool
+    {
+        return (($context instanceof QtiRunnerServiceContext) && $context->isAdaptive());
```

Comment:
No need for extra parentheses.

```suggestion
        return $context instanceof QtiRunnerServiceContext && $context->isAdaptive();
```